File: tests/cpp/cpu_save_state_replay_test.cpp
#include "machine/runtime/runtime.h"
#include "spec/blua32/instruction_format.h"
#include "spec/blua32/opcode.h"
#include "spec/bmsx/model.h"
#include "support/blua32_test_rom.h"

#include <array>
#include <iostream>
#include <stdexcept>

namespace {

void require(bool condition, const char* message) {
	if (!condition) throw std::runtime_error(message);
}

class SnapshotInput final : public bmsx::InputControllerInputSource {
public:
	void sampleInputControllerSnapshot(bmsx::InputControllerSnapshot&, bmsx::InputControllerSampleContext) override {}
	bool supervisorRequestLineHigh() const override { return false; }
	void applyInputControllerVibrationEffect(bmsx::i32, bmsx::f64, bmsx::f32) override {}
};

struct SnapshotMachine {
	bmsx::test::Blua32TestRom rom;
	SnapshotInput input;
	bmsx::Runtime runtime;

	explicit SnapshotMachine(const bmsx::test::Blua32TestImage& image)
		: rom(bmsx::test::encodeBlua32TestRom(bmsx::RomImageDomain::System, image))
		, runtime(bmsx::RuntimeOptions{rom.bytes, {}, bmsx::PSX_MACHINE_SPEC}, input) {
		runtime.machine.cpu.reset();
	}
};

bmsx::test::Blua32TestImage allocationImage() {
	bmsx::test::Blua32TestImage image;
	image.text.resize(8u * bmsx::INSTRUCTION_BYTES);
	std::span<bmsx::u8> code(image.text);
	bmsx::writeInstruction(code, 0, static_cast<bmsx::u8>(bmsx::OpCode::NEWT), 0, 0, 0);
	bmsx::writeInstruction(code, 1, static_cast<bmsx::u8>(bmsx::OpCode::WIDE), 0, 0, 0);
	bmsx::writeInstruction(code, 2, static_cast<bmsx::u8>(bmsx::OpCode::CLOSURE), 1, 0, 0);
	bmsx::writeInstruction(code, 3, static_cast<bmsx::u8>(bmsx::OpCode::WIDE), 0, 0, 0);
	bmsx::writeInstruction(code, 4, static_cast<bmsx::u8>(bmsx::OpCode::CLOSURE), 2, 0, 0);
	bmsx::writeInstruction(code, 5, static_cast<bmsx::u8>(bmsx::OpCode::RET), 0, 3, 0);
	bmsx::writeInstruction(code, 6, static_cast<bmsx::u8>(bmsx::OpCode::RET), 0, 0, 0);
	bmsx::writeInstruction(code, 7, static_cast<bmsx::u8>(bmsx::OpCode::RET), 0, 0, 0);
	image.functions = {
		{.firstWord = 0u, .wordCount = 6u, .maxStack = 3u},
		{.firstWord = 6u, .wordCount = 1u},
		{.firstWord = 7u, .wordCount = 1u, .staticClosure = false},
	};
	image.closureRelocations = {
		{2u, bmsx::test::blua32TestFunctionAddress(bmsx::RomImageDomain::System, 1u)},
		{4u, bmsx::test::blua32TestFunctionAddress(bmsx::RomImageDomain::System, 2u)},
	};
	return image;
}

void replayAllocationIdentities() {
	SnapshotMachine machine(allocationImage());
	auto& cpu = machine.runtime.machine.cpu;
	const auto anchor = cpu.captureRuntimeState();
	const auto strings = cpu.stringPool().captureState();
	require(cpu.runUntilDepth(0, 1000) == bmsx::RunResult::Halted, "program completes");
	const auto values = cpu.readCompletionValues();
	const std::array expectedIds{
		bmsx::asTable(values[0])->hashId,
		bmsx::asClosure(values[1])->hashId,
		bmsx::asClosure(values[2])->hashId,
	};
	const auto expected = cpu.captureRuntimeState();
	for (int pass = 0; pass < 4; ++pass) {
		cpu.stringPool().restoreState(strings);
		cpu.restoreRuntimeState(anchor);
		require(cpu.captureRuntimeState().nextObjectHashId == anchor.nextObjectHashId, "allocator sequence is restored");
		require(cpu.runUntilDepth(0, 1000) == bmsx::RunResult::Halted, "replay completes");
		const auto replay = cpu.readCompletionValues();
		require(bmsx::asTable(replay[0])->hashId == expectedIds[0], "table identity must replay");
		require(bmsx::asClosure(replay[1])->hashId == expectedIds[1], "cold canonical closure identity must replay");
		require(bmsx::asClosure(replay[2])->hashId == expectedIds[2], "dynamic closure identity must replay");
		const auto actual = cpu.captureRuntimeState();
		require(actual.nextObjectHashId == expected.nextObjectHashId, "next allocation matches");
		require(actual.luaHeap.trackedBytes == expected.luaHeap.trackedBytes, "replay heap accounting matches");
	}
}

void mirroredAllocationOrder() {
	auto image = allocationImage();
	image.functions[1].staticClosure = false;
	image.functions[1].upvalues = {{true, 0u}};
	SnapshotMachine machine(image);
	auto& cpu = machine.runtime.machine.cpu;
	require(cpu.stringPool().intern("error in error handling", false) == 2u, "mirrored protected-call error string id");
	require(cpu.stringPool().intern("Attempted to get length of an unsupported value.", false) == 3u, "mirrored fault error string id");
	cpu.runUntilDepth(0, 1000);
	const auto* closure = bmsx::asClosure(cpu.readCompletionValues()[1]);
	require(closure->upvalues[0]->hashId == closure->hashId + 1u, "closure allocation precedes its new upvalue");
}

void restoreAllocationWordWrap() {
	SnapshotMachine machine(allocationImage());
	auto& cpu = machine.runtime.machine.cpu;
	auto state = cpu.captureRuntimeState();
	state.nextObjectHashId = 0xffffffffu;
	cpu.restoreRuntimeState(state);
	require(cpu.createTable()->hashId == 0xffffffffu, "restored allocator retains the full u32 word");
	require(cpu.createTable()->hashId == 0u, "allocation sequence wraps as a u32 word");
	require(cpu.createTable()->hashId == 1u, "wrapped sequence advances deterministically");
}

void replayWeakCollectionSchedule() {
	SnapshotMachine machine(allocationImage());
	auto& cpu = machine.runtime.machine.cpu;
	const auto weakKey = cpu.stringPool().intern("weak");
	auto* weak = cpu.createTable(1, 0);
	auto* metatable = cpu.createTable(0, 1);
	metatable->setStringKey(cpu.stringPool().intern("__mode"), bmsx::valueString(cpu.stringPool().intern("v")));
	weak->metatable = metatable;
	weak->setInteger(1, bmsx::valueTable(cpu.createTable()));
	cpu.globals->setStringKey(weakKey, bmsx::valueTable(weak));
	for (int index = 0; index < 17; ++index) cpu.createTable(256, 0);
	const auto anchor = cpu.captureRuntimeState();
	const auto strings = cpu.stringPool().captureState();
	const auto allocateUntilCollected = [&]() {
		auto* current = bmsx::asTable(cpu.globals->getStringKey(weakKey));
		require(bmsx::valueIsTable(current->getInteger(1)), "restore must not collect a weak referent");
		for (int count = 1; count < 4096; ++count) {
			cpu.createTable(256, 0);
			if (bmsx::isNil(current->getInteger(1))) return count;
		}
		throw std::runtime_error("allocation did not reach the guest collection threshold");
	};
	const int expectedCount = allocateUntilCollected();
	const auto expected = cpu.captureRuntimeState();
	for (int pass = 0; pass < 3; ++pass) {
		cpu.stringPool().restoreState(strings);
		cpu.restoreRuntimeState(anchor);
		require(cpu.captureRuntimeState().luaHeap.trackedBytes == anchor.luaHeap.trackedBytes, "unreachable allocation debt survives restore");
		require(allocateUntilCollected() == expectedCount, "collection must happen after the same allocation");
		const auto actual = cpu.captureRuntimeState();
		require(actual.nextObjectHashId == expected.nextObjectHashId, "GC replay retains the allocator sequence");
		require(actual.luaHeap.trackedBytes == expected.luaHeap.trackedBytes, "GC recomputes actual live bytes after restore");
		require(actual.luaHeap.nextCollectionBytes == expected.luaHeap.nextCollectionBytes, "next GC threshold matches");
	}
}

void restoreUnrootedCanonicalClosure() {
	auto image = allocationImage();
	std::span<bmsx::u8> code(image.text);
	bmsx::writeInstruction(code, 0, static_cast<bmsx::u8>(bmsx::OpCode::WIDE), 0, 0, 0);
	bmsx::writeInstruction(code, 1, static_cast<bmsx::u8>(bmsx::OpCode::CLOSURE), 0, 0, 0);
	bmsx::writeInstruction(code, 2, static_cast<bmsx::u8>(bmsx::OpCode::KNIL), 0, 0, 0);
	bmsx::writeInstruction(code, 3, static_cast<bmsx::u8>(bmsx::OpCode::NEWT), 0, 0, 0);
	bmsx::writeInstruction(code, 4, static_cast<bmsx::u8>(bmsx::OpCode::RET), 0, 1, 0);
	const auto address = bmsx::test::blua32TestFunctionAddress(bmsx::RomImageDomain::System, 1u);
	image.closureRelocations = {{1u, address}};
	SnapshotMachine machine(image);
	auto& cpu = machine.runtime.machine.cpu;
	const auto cold = cpu.captureRuntimeState();
	cpu.runUntilDepth(0, 2);
	const auto warm = cpu.captureRuntimeState();
	require(warm.frames[0].registers[0].tag == bmsx::CpuValueStateTag::Nil, "static closure is no longer a guest root");
	bool captured = false;
	for (const auto& object : warm.objects) {
		if (object.kind == bmsx::CpuObjectState::Kind::Closure && object.closureCanonical && object.functionAddress == address) captured = true;
	}
	require(captured, "snapshot must retain unrooted canonical cache entries");
	cpu.runUntilDepth(0, 1000);
	const auto expectedId = bmsx::asTable(cpu.readCompletionValues()[0])->hashId;
	cpu.restoreRuntimeState(cold);
	cpu.restoreRuntimeState(warm);
	cpu.runUntilDepth(0, 1000);
	require(bmsx::asTable(cpu.readCompletionValues()[0])->hashId == expectedId, "warm cache restore must preserve the next guest allocation identity");
}

void captureDoesNotSynchronizeGlobals() {
	auto image = allocationImage();
	image.globalNames = {"answer"};
	std::span<bmsx::u8> code(image.text);
	bmsx::writeInstruction(code, 0, static_cast<bmsx::u8>(bmsx::OpCode::K1), 0, 0, 0);
	bmsx::writeInstruction(code, 1, static_cast<bmsx::u8>(bmsx::OpCode::SETGL), 0, 0, 0);
	bmsx::writeInstruction(code, 2, static_cast<bmsx::u8>(bmsx::OpCode::RET), 0, 0, 0);
	image.closureRelocations.clear();
	SnapshotMachine machine(image);
	auto& cpu = machine.runtime.machine.cpu;
	cpu.runUntilDepth(0, 1000);
	const auto before = cpu.globals->captureRuntimeState();
	const auto snapshot = cpu.captureRuntimeState();
	require(cpu.globals->captureRuntimeState().hash.size() == before.hash.size(), "capture must not mutate global table storage");
	cpu.restoreRuntimeState(snapshot);
	require(cpu.globals->captureRuntimeState().hash.size() == before.hash.size(), "restore retains backing table capacity");
	require(bmsx::asNumber(cpu.getGlobalByKey(cpu.stringPool().intern("answer"))) == 1.0, "restore retains the newer registerfile value independently");
}

void restoreHardHalt() {
	auto image = allocationImage();
	image.closureRelocations.clear();
	std::span<bmsx::u8> code(image.text);
	bmsx::writeInstruction(code, 0, static_cast<bmsx::u8>(bmsx::OpCode::WIDE), 0, 0, 0);
	bmsx::writeInstruction(code, 1, static_cast<bmsx::u8>(bmsx::OpCode::WIDE), 0, 0, 0);
	SnapshotMachine machine(image);
	auto& cpu = machine.runtime.machine.cpu;
	const auto start = cpu.captureRuntimeState();
	cpu.runUntilDepth(0, 1000);
	const auto halted = cpu.captureRuntimeState();
	require(halted.hardHalted, "invalid WIDE body latches hard halt");
	cpu.restoreRuntimeState(start);
	require(!cpu.captureRuntimeState().hardHalted, "restoring before halt clears the latch");
	cpu.restoreRuntimeState(halted);
	cpu.runUntilDepth(0, 1000);
	const auto after = cpu.captureRuntimeState();
	require(after.hardHalted && after.frames[0].pc == halted.frames[0].pc, "restored hard halt does not execute further instructions");
}

} // namespace

int main() {
	try {
		replayAllocationIdentities();
		mirroredAllocationOrder();
		restoreAllocationWordWrap();
		replayWeakCollectionSchedule();
		restoreUnrootedCanonicalClosure();
		captureDoesNotSynchronizeGlobals();
		restoreHardHalt();
		std::cout << "CPU save-state replay tests passed\n";
		return 0;
	} catch (const std::exception& error) {
		std::cerr << error.what() << '\n';
		return 1;
	}
}

File: tests/lua/cpu_save_state_replay.test.ts
import assert from 'node:assert/strict';
import { test } from 'node:test';
import { RunResult } from '../../machine/ts/machine/cpu/cpu';
import { Table } from '../../machine/ts/machine/cpu/table';
import { valueString } from '../../machine/ts/machine/cpu/value';
import { INSTRUCTION_BYTES, writeInstruction } from '../../machine/ts/spec/blua32/instruction_format';
import { SYSTEM_ROM_BASE } from '../../machine/ts/spec/bmsx/memory_map';
import { OpCode } from '../../machine/ts/spec/blua32/opcode';
import { blua32TestFunctionAddress, createTestSystemCpu, linkRawTestSystemBlua32, linkTestSystemBlua32 } from '../helpers/blua32';
import { compileLuaSource } from './cpu_test_harness';

test('the mirrored CPU allocates a closure before its new upvalues and interns the same error prefix', () => {
	const { cpu } = createTestSystemCpu(linkTestSystemBlua32(compileLuaSource(`
local root = {}
return function() return root end
`, 'snapshot-allocation-order.lua')));
	assert.equal(cpu.stringPool.find('error in error handling'), 2);
	assert.equal(cpu.stringPool.find('Attempted to get length of an unsupported value.'), 3);
	cpu.runUntilDepth(0, 1000);
	const state = cpu.captureRuntimeState();
	const result = state.completionValues[0];
	assert.ok(result.tag === 'closure');
	const closure = state.objects[result.id];
	assert.ok(closure.kind === 'closure');
	assert.equal(state.objects[closure.upvalues[0]].hashId, closure.hashId + 1);
});

test('repeated restore replays table and cold canonical/dynamic closure identities', () => {
	const image = linkTestSystemBlua32(compileLuaSource(`
local cached<const> = function() return 1 end
local dynamic = function() return 2 end
return {}, cached, dynamic
`, 'snapshot-identities.lua'));
	const { cpu } = createTestSystemCpu(image);
	const anchor = cpu.captureRuntimeState();
	const strings = cpu.stringPool.captureState();
	assert.equal(cpu.runUntilDepth(0, 1000), RunResult.Halted);
	const expected = cpu.captureRuntimeState();
	for (let pass = 0; pass < 4; pass += 1) {
		cpu.stringPool.restoreState(strings);
		cpu.restoreRuntimeState(anchor);
		assert.deepEqual(cpu.captureRuntimeState(), anchor);
		assert.equal(cpu.runUntilDepth(0, 1000), RunResult.Halted);
		assert.deepEqual(cpu.captureRuntimeState(), expected);
	}
});

test('restored allocation words wrap identically to the native u32 allocator', () => {
	const { cpu } = createTestSystemCpu(linkTestSystemBlua32(compileLuaSource('return 0', 'snapshot-id-wrap.lua')));
	const state = cpu.captureRuntimeState();
	state.nextObjectHashId = 0xffffffff;
	cpu.restoreRuntimeState(state);
	assert.equal(cpu.createTable().hashId, 0xffffffff);
	assert.equal(cpu.createTable().hashId, 0);
	assert.equal(cpu.createTable().hashId, 1);
});

test('checkpoint capture preserves global slot/backing-table state without synchronizing it', () => {
	const code = new Uint8Array(3 * INSTRUCTION_BYTES);
	writeInstruction(code, 0, OpCode.K1, 0, 0, 0);
	writeInstruction(code, 1, OpCode.SETGL, 0, 0, 0);
	writeInstruction(code, 2, OpCode.RET, 0, 0, 0);
	const { cpu } = createTestSystemCpu(linkRawTestSystemBlua32({
		text: code,
		functions: [{ firstWord: 0, wordCount: 3 }],
		globalNames: ['answer'],
	}));
	assert.equal(cpu.runUntilDepth(0, 1000), RunResult.Halted);
	const storage = cpu.globals.captureRuntimeState();
	const heap = cpu.luaHeap.captureState();
	const snapshot = cpu.captureRuntimeState();
	assert.deepEqual(cpu.globals.captureRuntimeState(), storage);
	assert.deepEqual(cpu.luaHeap.captureState(), heap);
	assert.equal(cpu.getGlobalByKey(cpu.stringPool.find('answer')!), 1);
	cpu.restoreRuntimeState(snapshot);
	assert.deepEqual(cpu.captureRuntimeState(), snapshot);
	assert.deepEqual(cpu.globals.captureRuntimeState(), storage);
	assert.equal(cpu.getGlobalByKey(cpu.stringPool.find('answer')!), 1);
});

test('restore retains weak referents until the original allocation-triggered collection', () => {
	const { cpu } = createTestSystemCpu(linkTestSystemBlua32(
		compileLuaSource('return 0', 'snapshot-weak.lua'),
	));
	const weakKey = cpu.stringPool.intern('weak');
	const weak = cpu.createTable(1, 0);
	const metatable = cpu.createTable(0, 1);
	metatable.setStringKey(cpu.stringPool.intern('__mode'), valueString(cpu.stringPool.intern('v')));
	weak.metatable = metatable;
	weak.set(1, cpu.createTable());
	cpu.globals.setStringKey(weakKey, weak);
	// These allocations have left guest accounting debt but are unreachable.
	for (let index = 0; index < 17; index += 1) cpu.createTable(256, 0);
	const anchor = cpu.captureRuntimeState();
	const strings = cpu.stringPool.captureState();
	const allocateUntilCollected = (): number => {
		const current = cpu.globals.getStringKey(weakKey) as Table;
		assert.ok(current.get(1) instanceof Table);
		for (let count = 1; count < 4096; count += 1) {
			cpu.createTable(256, 0);
			if (current.get(1) === null) return count;
		}
		assert.fail('allocation did not reach the guest collection threshold');
	};
	const expectedCount = allocateUntilCollected();
	const expected = cpu.captureRuntimeState();
	for (let pass = 0; pass < 3; pass += 1) {
		cpu.stringPool.restoreState(strings);
		cpu.restoreRuntimeState(anchor);
		assert.deepEqual(cpu.captureRuntimeState(), anchor);
		assert.equal(allocateUntilCollected(), expectedCount);
		assert.deepEqual(cpu.captureRuntimeState(), expected);
	}
});

test('a materialized but unrooted canonical closure remains part of the checkpoint', () => {
	const code = new Uint8Array(6 * INSTRUCTION_BYTES);
	writeInstruction(code, 0, OpCode.WIDE, 0, 0, 0);
	writeInstruction(code, 1, OpCode.CLOSURE, 0, 0, 0);
	writeInstruction(code, 2, OpCode.KNIL, 0, 0, 0);
	writeInstruction(code, 3, OpCode.NEWT, 0, 0, 0);
	writeInstruction(code, 4, OpCode.RET, 0, 1, 0);
	writeInstruction(code, 5, OpCode.RET, 0, 0, 0);
	const address = blua32TestFunctionAddress(SYSTEM_ROM_BASE, 1);
	const { cpu } = createTestSystemCpu(linkRawTestSystemBlua32({
		text: code,
		functions: [{ firstWord: 0, wordCount: 5 }, { firstWord: 5, wordCount: 1 }],
		closureRelocations: [{ wordIndex: 1, functionAddress: address }],
	}));
	const cold = cpu.captureRuntimeState();
	assert.equal(cpu.runUntilDepth(0, 2), RunResult.Yielded);
	const warm = cpu.captureRuntimeState();
	assert.deepEqual(warm.frames[0].registers, [{ tag: 'nil' }]);
	assert.ok(warm.objects.some(object => object.kind === 'closure'
		&& object.canonical && object.functionAddress === address));
	cpu.runUntilDepth(0, 1000);
	const expected = cpu.captureRuntimeState();
	cpu.restoreRuntimeState(cold);
	cpu.restoreRuntimeState(warm);
	assert.deepEqual(cpu.captureRuntimeState(), warm);
	cpu.runUntilDepth(0, 1000);
	assert.deepEqual(cpu.captureRuntimeState(), expected);
});

test('hard-halt is restored independently of the live future CPU latch', () => {
	const code = new Uint8Array(3 * INSTRUCTION_BYTES);
	writeInstruction(code, 0, OpCode.WIDE, 0, 0, 0);
	writeInstruction(code, 1, OpCode.WIDE, 0, 0, 0);
	writeInstruction(code, 2, OpCode.RET, 0, 0, 0);
	const { cpu } = createTestSystemCpu(linkRawTestSystemBlua32({
		text: code,
		functions: [{ firstWord: 0, wordCount: 3 }],
	}));
	const start = cpu.captureRuntimeState();
	cpu.runUntilDepth(0, 1000);
	const halted = cpu.captureRuntimeState();
	assert.equal(halted.hardHalted, true);
	cpu.restoreRuntimeState(start);
	assert.equal(cpu.captureRuntimeState().hardHalted, false);
	cpu.restoreRuntimeState(halted);
	cpu.runUntilDepth(0, 1000);
	const after = cpu.captureRuntimeState();
	assert.equal(after.hardHalted, true);
	assert.deepEqual(after.frames, halted.frames);
});
